extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcRecorder.java

```diff
@@ -498,4 +508,67 @@ public void accept(RoutingContext routingContext) {
             }
         };
     }
+
+    public Supplier<TenantIdentityProvider> createTenantIdentityProvider(String tenantName) {
+        return new Supplier<TenantIdentityProvider>() {
+            @Override
+            public TenantIdentityProvider get() {
+                return new TenantSpecificOidcIdentityProvider(tenantName);
+            }
+        };
+    }
+
+    private static final class TenantSpecificOidcIdentityProvider extends OidcIdentityProvider
+            implements TenantIdentityProvider {
+
+        private final String tenantId;
+        private final BlockingSecurityExecutor blockingExecutor;
+
+        private TenantSpecificOidcIdentityProvider(String tenantId) {
+            super(Arc.container().instance(DefaultTenantConfigResolver.class).get(),
+                    Arc.container().instance(BlockingSecurityExecutor.class).get());
+            this.blockingExecutor = Arc.container().instance(BlockingSecurityExecutor.class).get();
+            if (tenantId.equals(DEFAULT_TENANT_ID)) {
+                OidcConfig config = Arc.container().instance(OidcConfig.class).get();
+                this.tenantId = config.defaultTenant.getTenantId().orElse(OidcUtils.DEFAULT_TENANT_ID);
+            } else {
+                this.tenantId = tenantId;
+            }
+        }
+
+        @Override
+        public Uni<SecurityIdentity> authenticate(AccessTokenCredential token) {
+            return authenticate(new TokenAuthenticationRequest(token));
+        }
+
+        @Override
+        protected Uni<TenantConfigContext> resolveTenantConfigContext(TokenAuthenticationRequest request,
+                AuthenticationRequestContext context) {
+            return tenantResolver.resolveContext(tenantId).onItem().ifNull().failWith(new Supplier<Throwable>() {
+                @Override
+                public Throwable get() {
+                    return new AuthenticationFailedException(
```

Comment:
I may have some cases where some of such details may be reported too which will need to be reviewed, as now users are very concerned about all sort of information leaks, so it is only here that I'm trying to focus :-)

Reply:
@sberyozkin please let me divide response to this into 2 parts:

- I'm fine with the change you propose
- I don't understand how could there be any information leak at all. We don't append `AuthenticationFailedException` body into response, I just checked `DefaultAuthFailureHandler` and `QuarkusErrorHandler`. Maybe OIDC does it somewhere? If so, that is what I consider issue, not informative exception message. It will be real fun to debug it without information for which tenant this failed. To my knowledge, only way this auth exception message is going to be showed is when user dfines exception mapper / failure handler. Maybe I am wrong?

Reply:
I've used `OIDCException` and with less details now.